### PR TITLE
Getting rid of lock().unwrap()

### PR DESCRIPTION
**Pull request by dr-orlovsky**
PR removes 16 (!) `unwraps()` which are unnecessary and were already covered by an error case.

This is quite important for the overall library stability

### CODE REVIEW

File: src/types.rs
Comment (afilini) [inline src/types.rs:330]: I wouldn't use this error here, because `CouldntLockReader` has a very specific meaning:

> Couldn't take a lock on the reader mutex. This means that there's already another reader thread running

If a `PoisonError` happens there's not much the caller can do to recover, it has to throw away the client and recreate it. I would add an error variant that reflects this better, and write some documentation to let users know what to do in case that error happens.

File: src/stream.rs
Comment (afilini) [inline src/stream.rs:11]: I agree that `BrokenPipe` is the error variant that more closely maps to a `PoisonError`, but I would at least log what's happening here first before returning, to avoid confusing the user.

Same for `write()` and `flush()` below

File: src/raw_client.rs
Comment (afilini) [inline src/raw_client.rs:474]: If we are unable to wake up the thread I guess we should try with all the others before giving up, otherwise all the other threads that are still alive would remain hanging forever waiting for their response.

**Review by afilini — CHANGES_REQUESTED** [review]

File: src/types.rs
Comment (dr-orlovsky) [reply src/types.rs:330]: I am strongly against panics at runtime which happens not due to app design bugs, especially when you can avoid them in a small cost. What client app can do is to handle the error, terminate the tread gracefully (save the state, notify users, write debug logs, send debug info to developer) and restart the thread.

File: src/stream.rs
Comment (dr-orlovsky) [reply src/stream.rs:11]: Agree, will update

File: src/raw_client.rs
Comment (dr-orlovsky) [reply src/raw_client.rs:474]: Agree, will update

File: src/types.rs
Comment (afilini) [reply src/types.rs:330]: Yeah I was suggesting to add a specific variant that reflects the true meaning of the error, and then the user can choose how to handle that variant specifically.

I generally agree with you that it's better to return an error when possible, but on the other end it's very important to ensure that the code can't accidentally end up in an inconsistent/unexpected state: in that regard, the panics on `lock()` were used to simplify things a little bit.

Basically we chose to panic rather than potentially incur in some kind of logic bug, which I still think it's a better trade-off in many cases.

Comment (dr-orlovsky) [reply src/types.rs:330]: I am ok with adding error-specific variant. Or you are against removing `unwrap()`s on `lock()`?

**Comment by dr-orlovsky** [comment]
If possible will prefer adding error-specific variants after #36 merge, so I can rebase this PR on it and add error displays at the same time

File: src/types.rs
Comment (afilini) [reply src/types.rs:330]: No that's fine, I'm just saying that now that we don't panic we have to be a bit more careful and think about the potential edge cases..

**Comment by dr-orlovsky** [comment]
@afilini rebased & fixed according to your comments

**Review by afilini — APPROVED** [review]

**Pull request merged**
